### PR TITLE
feat: make http client retryable

### LINKED ISSUE #6600: Support retry in cli command
# Summary

Support retry in cli commands

# Motivation

CLI may fail to establish connection to server when the network is unreliable. It would be nice to have retry options on temporary failures.

# Proposal

Add option `--retry-limit=3` to let client retry on specific failures, e.g. `Failed to establish connection to argocd-server:443:`

### PR DESCRIPTION
**Pull request by yujunz**
Closes #6600

### CODE REVIEW

File: cmd/argocd/commands/root.go [hunk 1]
```diff
@@ -69,5 +69,6 @@ func NewCommand() *cobra.Command {
 	command.PersistentFlags().StringSliceVarP(&clientOpts.Headers, "header", "H", []string{}, "Sets additional header to all requests made by Argo CD CLI. (Can be repeated multiple times to add multiple headers, also supports comma separated headers)")
 	command.PersistentFlags().BoolVar(&clientOpts.PortForward, "port-forward", config.GetBoolFlag("port-forward"), "Connect to a random argocd-server port using port forwarding")
 	command.PersistentFlags().StringVar(&clientOpts.PortForwardNamespace, "port-forward-namespace", config.GetFlag("port-forward-namespace", ""), "Namespace name which should be used for port forwarding")
+	command.PersistentFlags().IntVar(&clientOpts.RetryMax, "retry-max", 0, "Maximum number of retries")
```
Comment (jannfis) [inline cmd/argocd/commands/root.go:72]: I don't think `retry-max` is a good name for the option. The name should imply _what_ is actually retried, although I cannot think of a _proper_ name right now. Maybe `http-retries` or `api-retries` would fit better? Also, the description should be adapted to be a little more verbose.

**Review by jannfis — CHANGES_REQUESTED** [review]
Thanks for this PR @yujunz. 

LGTM and non-intrusive to me, and I think it's a really useful feature to have to improve stability of pipelines. However, I have some usability concern, please see below.

Also, since you are introducing a new dependency, can you please give some information about that and its license? Thanks.

**Comment by yujunz** [comment]
> Also, since you are introducing a new dependency, can you please give some information about that and its license? Thanks.

The new dependency hashicorp/go-retryablehttp is released under [MPL 2.0](https://github.com/hashicorp/go-retryablehttp/blob/master/LICENSE) which should be compatible to compile into Apache licensed code like argocd if I understand the [FAQ](https://www.mozilla.org/en-US/MPL/2.0/FAQ/) correctly.

> Q13: May I combine MPL-licensed code and BSD-licensed code in the same executable program? What about Apache?
> Yes to both. Mozilla currently does this with BSD-licensed code. For example, libvpx, which is used in Firefox to decode WebM video, is under a BSD license.

File: cmd/argocd/commands/root.go (hunk 1 above)
Comment (yujunz) [reply cmd/argocd/commands/root.go]: Renamed to `http-retry-max` and adapted the description to be more verbose.

**Review by jannfis — APPROVED** [review]
LGTM, thank you!

**Pull request merged**